blocks/main_loop.py

```diff
@@ -246,13 +248,11 @@ def _check_finish_training(self, level):
         # the iteration the corresponding log record can be found only in
         # the previous row.
         if (self.log.current_row.training_finish_requested or
-                self.log.current_row.batch_interrupt_received or
-                self.log.previous_row.batch_interrupt_received):
-            raise TrainingFinish
+            self.status._batch_interrupt_received):
```

Comment:
I see you ran into a nasty part of PEP8. This should work:
```python
if (self.log.current_row.training_finish_requested or
        self.status._batch_interrupt_received):
    raise TrainingFinish
```
The indentation should be allowed like this in order to avoid `self.status` starting at the same point as `raise`.